File: code/ryzom/client/src/r2/r2_config.cpp
// Ryzom - MMORPG Framework <http://dev.ryzom.com/projects/ryzom/>
// Copyright (C) 2010  Winch Gate Property Limited
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

#include "stdpch.h"
#include "r2_config.h"


using namespace NLMISC;

namespace R2
{

CConfigVarRGBA	CV_MapEntityHighlightColor("MapEntityHighlightColor", CRGBA(255, 255, 255, 127));
CConfigVarRGBA	CV_MapEntitySelectColor("MapEntitySelectColor", CRGBA(255, 0, 0, 255));
CConfigVarRGBA	CV_MapEntityFrozenColor("MapEntityFrozenColor", CRGBA(0,  255,  255,  80));
CConfigVarRGBA	CV_MapEntityLockedColor("MapEntityLockedColor", CRGBA(127,  0,  200,  80));
CConfigVarRGBA	CV_ArrayInstanceColor("ArrayInstanceColor", CRGBA(255, 255, 255, 127));
CConfigVarString CV_MapEntitySelectTexture("MapEntitySelectTexture", "r2_icon_select.tga");
CConfigVarString CV_MapEntityDefaultTexture("MapEntityDefaultTexture", "brick_default.tga");
CConfigVarString CV_MapEntityFarTexture("MapEntityFarTexture", "r2_icon_far.tga");
CConfigVarFloat  CV_MapEntityFarArrowSize("MapEntityFarArrowSize", 10.f);
CConfigVarString CV_MapEntitySmallTexture("MapEntitySmallTexture", "r2_icon_map_entity_small.tga");
CConfigVarString CV_MapEntitySmallHighlightTexture("MapEntitySmallHighlightTexture", "r2_icon_map_entity_small_highlight.tga");
CConfigVarString CV_MapEntityOrientTexture("MapEntityOrientTexture", "r2_icon_map_entity_orient.tga");
CConfigVarFloat  CV_MapEntityOrientOriginDist("MapEntityOrientOriginDist", 10.f);
CConfigVarFloat  CV_MapEntityOrientBlendTimeInMs("MapEntityOrientBlendTimeInMs", 300.f);
CConfigVarFloat  CV_MapEntityOrientOriginDistSmall("MapEntityOrientOriginDistSmall", 8.f);
CConfigVarFloat  CV_MapEntityCloseDist("MapEntityCloseDist", 4.0f);

//
CConfigVarString CV_MapGlowStarTexture("MapGlowStarTexture", "r2_glow_star.tga");
CConfigVarFloat  CV_MapGlowStarSize("MapGlowStarSize", 10.f);
CConfigVarFloat	CV_MapGlowStarSpeed[2] =
{
	CConfigVarFloat("MapGlowStarSpeed1", 0.50f),
	CConfigVarFloat("MapGlowStarSpeed2", -0.60f)
};
//
CConfigVarFloat CV_FloatingShapeRefScale("FloatingShapeRefScale", 1.f);
CConfigVarFloat CV_RegionFadeTimeInMs("RegionFadeTimeInMs", 300.f);
//
CConfigVarRGBA CV_FocusedRegionColor("FocusedRegionColor", CRGBA(63,  127,  255,  100));
CConfigVarRGBA CV_SelectedRegionColor("SelectedRegionColor", CRGBA(192,  127,  64,  100));
CConfigVarRGBA CV_UnselectedRegionColor("UnselectedRegionColor", CRGBA(0,  0,  255,  80));
CConfigVarRGBA CV_FrozenRegionColor("FrozenRegionColor", CRGBA(0,  255,  255,  80));
CConfigVarRGBA CV_LockedRegionColor("LockedRegionColor", CRGBA(127,  0,  200,  80));
//
CConfigVarRGBA CV_FocusedInstanceColor("FocusedInstanceColor", CRGBA(200,  32,  64, 127));
CConfigVarRGBA CV_SelectedInstanceColor("SelectedInstanceColor", CRGBA(127,  127,  127));
CConfigVarRGBA CV_UnselectedInstanceColor("UnselectedInstanceColor", CRGBA(0,  0,  0));
CConfigVarRGBA CV_FrozenInstanceColor("FrozenInstanceColor", CRGBA(0,  255,  255,  80));
CConfigVarRGBA CV_LockedInstanceColor("LockedInstanceColor", CRGBA(127,  0,  200,  80));
//
CConfigVarSInt32 CV_MapAutoPanBorder("MapAutoPanBorder", 10);
CConfigVarSInt32 CV_MapAutoPanDeltaInMs("MapAutoPanDeltaInMs", 300);
CConfigVarSInt32 CV_MapAutoFastPanDeltaInMs("MapAutoPanFastDeltaInMs", 150);
CConfigVarSInt32 CV_MapAutoPanSpeedInPixels("MapAutoPanSpeedInPixels", 10);
CConfigVarSInt32 CV_MapAutoFastPanNumTicks("MapAutoFastPanNumTicks", 6);
//
CConfigVarString	CV_FootStepMapTexture("FootStepMapTexture", "r2_map_foot_steps.tga");
CConfigVarString	CV_FootStepDecalTexture("FootStepDecalTexture", "r2_foot_steps.tga");
CConfigVarFloat		CV_FootStepMapWidth("FootStepMapWidth", 3.f);
CConfigVarFloat		CV_FootStepDecalUScale("FootStepDecalUScale", 1.5f);
CConfigVarFloat		CV_FootStepDecalWidth("FootStepDecalWidth", 0.15f);
CConfigVarString	CV_WanderDecalTexture("WanderDecalTexture", "r2_wander.tga");
CConfigVarFloat		CV_WanderDecalSize("WanderDecalSize", 0.55f);
CConfigVarRGBA		CV_FootStepDecalSelectedColor("FootStepSelectedColor", CRGBA(255, 255, 255, 255));
CConfigVarRGBA		CV_FootStepDecalHiddenColor("FootStepHiddenColor", CRGBA(0, 0, 255, 80));
CConfigVarRGBA		CV_FootStepDecalFocusedColor("FootStepFocusedColor", CRGBA(255, 255, 255, 127));
CConfigVarRGBA		CV_FootStepMapSelectedColor("FootStepMapSelectedColor", CRGBA(255, 255, 255, 255));
CConfigVarRGBA		CV_FootStepMapHiddenColor("FootStepMapHiddenColor", CRGBA(0, 0, 255, 80));
CConfigVarRGBA		CV_FootStepMapFocusedColor("FootStepMapFocusedColor", CRGBA(255, 255, 255, 127));
//
CConfigVarFloat		CV_AutoGroupMaxDist("AutoGroupMaxDist", 5.f);
//
CConfigVarRGBA		CV_InaccessiblePosColor0("InaccessiblePosColor0", CRGBA(255, 0, 0, 255));
CConfigVarRGBA		CV_InaccessiblePosColor1("InaccessiblePosColor1", CRGBA(200, 217, 0, 255));
CConfigVarFloat		CV_InaccessiblePosAnimDurationInMS("InaccessiblePosAnimDurationInMS", 500);
//
CConfigVarFloat		CV_DecalTopBlendStartDist("DecalTopBlendStartDist", 3.f);
CConfigVarFloat		CV_DecalBottomBlendStartDist("DecalBottomBlendStartDist", 1.f);
CConfigVarFloat		CV_DecalBlendLength("DecalBlendLength", 1.5f);

CConfigVarString	CV_MapEntityInvalidTexture("MapEntityInvalidTexture", "r2_icon_map_invalid.tga");
CConfigVarString	CV_MapEntityInvalidTextureSmall("MapEntityInvalidTextureSmall", "r2_icon_map_invalid_small.tga");



} // R2

File: ryzom/common/data_common/r2/r2_config.lua
-- config file for the r2 environment


verboseDebugInfo = false       -- when set to true, debug infos will be preceded by the number of the line that generated them
traceFunctions   = false      -- when set to true, each function enter / function leave are dumped to the stack


-- TODO nico : a table with predefined colors
r2.Config = 
{	
	PrimDisplayEnabled		= true,	     -- should route and zones be displayed ?
	FloatingShapeRefScale	= 1.0,       -- size of the floating mesh displayed at the mouse position when view-ray didn't hit a valid pos (in creation mode)
	RegionFadeTimeInMs		= 300,       -- time in ms needed for a region to get higlighted
	TestPaletteSelection    = false,     -- complete the palette for a test for selection by ecosystem & level	
	VerboseGetUI            = false,     -- some getUI can be made silent when they fail by turning this option to 'false'
	ResetWindowPos			= true,		 -- TMP : should windows pos be reseted at startup ?
	ActMaxQuota				= 99,        -- 'hardcoded" : max content that can be added into an act (checked by server too, but value is not taken from this file of course ..)
	-- decals look	
	HightlightDecalLook = 
	{
		DurationInMs = 1000,
		Texture = "r2_highlight_decal.tga",
		EndScaleFactor = 1,
		EndAngleInDegrees = 0,
		StartDiffuse = CRGBA(255, 255, 255),
		EndDiffuse = CRGBA(127, 0, 255),
		StartEmissive = CRGBA(0, 0, 0),
		EndEmissive = CRGBA(0, 0, 0),
	},
	SelectDecalLook = 
	{
		DurationInMs = 1000,
		Texture = "r2_select_decal.tga",
		EndScaleFactor = 1.2,
		EndAngleInDegrees = 0,
		StartDiffuse = CRGBA(255, 0, 0),
		EndDiffuse = CRGBA(255, 0, 0, 127),
		StartEmissive = CRGBA(0, 0, 0),
		EndEmissive = CRGBA(127, 127, 127),
	},
	SelectingDecalLook = 
	{
		DurationInMs = 200,
		Texture = "r2_select_decal.tga",
		EndScaleFactor = 1.2,
		EndAngleInDegrees = 90,
		StartDiffuse = CRGBA(255, 255, 255),
		EndDiffuse = CRGBA(255, 255, 255, 0),
		StartEmissive = CRGBA(0, 0, 0),
		EndEmissive = CRGBA(0, 0, 0)
	},
	PionneerDecalLook = 
	{
		DurationInMs = 3000,
		Texture = "r2_pionneer_edit.tga",
		EndScaleFactor = 1.05,
		EndAngleInDegrees = 0,
		StartDiffuse = CRGBA(255, 0, 0, 255),
		EndDiffuse = CRGBA(255, 0, 0, 192),
		StartEmissive = CRGBA(0, 64, 0),
		EndEmissive = CRGBA(0, 0, 0)
	},
	--
	MapEntityInvalidTexture = "r2_icon_map_invalid.tga",		   -- over displayed on minimap when object in on invalid pos
	MapEntityInvalidTextureSmall = "r2_icon_map_invalid_small.tga", -- over displayed on minimap when object in on invalid pos, small version
	MapEntityFrozenColor = CRGBA(0,  255,  255,  80),
	MapEntityLockedColor = CRGBA(127,  0,  200,  80),
	ArrayInstanceColor = CRGBA(255, 255, 255, 127),
	MapEntityDefaultTexture  = "brick_default.tga";        -- Bitmap to be used when one of the following bitmaps wasn't found
	MapEntitySelectTexture   = "r2_icon_select.tga",       -- The 'circle' bitmap draw on the map over selected / highlighted entities
	MapEntityHighlightColor  = CRGBA(255, 255, 255, 255),      -- Color of selection bitmap over an entity on the map
	MapEntitySelectColor     = CRGBA(255, 0, 0, 255),  -- Color of highlight bitmap over an entity on the map
	MapEntityFarTexture      = "r2_icon_far.tga",          -- Texture to display on the border of the map when the current selection is out 
	                                                       -- ... of the current visible world part in the map
	MapEntityFarArrowSize	 = 10,                         -- Size of the "MapEntityFarTexture" texture in pixels (scale applied ..)
	MapEntitySmallTexture    = "r2_icon_map_entity_small.tga",    -- Texture for entities when map is zoomed out
	MapEntitySmallHighlightTexture = "r2_icon_map_entity_small_highlight.tga",
	MapEntityOrientTexture   = "r2_icon_map_entity_orient.tga", -- arrow to signal entity orientation in close view
	MapEntityOrientOriginDist = 9,							-- distance  from entity to its orient texture
	MapEntityOrientOriginDistSmall = 5,					   -- distance  from entity to its orient texture when the entity is small and wihle it is being rotated
	MapEntityOrientBlendTimeInMs = 300,					   -- time that the orientation arrow need to get back to alpha = 0 when rotation is stopped		
	MapEntityCloseDist		 = 0.24,					   -- number of meter per pixel for the close view
	MapGlowStarTexture		 = "r2_glow_star.tga",         -- Texture displayed in the border of the map to signal that an object is too far to be seen
	MapGlowStarSpeed1		 = 0.50,                       -- rotation speed of first  map glow star
	MapGlowStarSpeed2		 = -0.60,                      -- rotation speed of second map glow star
	MapGlowStarSize          = 7,                           -- size of glow star 	
	-- color of regions in the 3D scene
	FocusedRegionColor = CRGBA(63,  127,  255,  100),
	SelectedRegionColor = CRGBA(192,  127,  64,  100),
	UnselectedRegionColor = CRGBA(0,  0,  255,  80),
	FrozenRegionColor = CRGBA(0,  255,  255,  80),
	LockedRegionColor = CRGBA(127,  0,  200,  80),
	-- colors for selected / focused entities / instance in scene (such as creature & npcs)
	FocusedInstanceColor = CRGBA(200,  32,  64, 127),
	SelectedInstanceColor = CRGBA(127,  127,  127),
	UnselectedInstanceColor = CRGBA(0,  0,  0),
	FrozenInstanceColor = CRGBA(0,  255,  255,  80),
	LockedInstanceColor = CRGBA(127,  0,  200,  80),
	-- world map auto pan
	MapAutoPanBorder = 20,  -- number of pixels in the border of the map for which the auto pan is tested
	MapAutoPanDeltaInMs = 300,		-- delta between each map pan when the mouse is on an auto-pan region
	MapAutoPanFastDeltaInMs = 100,  -- delta between each map pan when fast pan is active
	MapAutoPanSpeedInPixels = 15,   -- number of pixel to pan the map
	MapAutoFastPanNumTicks = 4,     -- number of 'slow' pan to do before entering fast pan mode
	-- foot steps / wander look
	FootStepMapTexture = "r2_map_foot_steps.tga",	
	FootStepDecalTexture = "r2_foot_steps.tga",
	FootStepMapWidth = 3,
	FootStepDecalUScale = 1.5,
	FootStepDecalWidth = 0.15,	
	FootStepDecalSelectedColor = CRGBA(255, 255, 255, 255),
	FootStepDecalHiddenColor = CRGBA(0, 0, 255, 120),
	FootStepDecalFocusedColor = CRGBA(255, 255, 255, 127),
	FootStepMapSelectedColor = CRGBA(180, 0, 192, 255),
	FootStepMapHiddenColor = CRGBA(150, 0, 192, 100),
	FootStepMapFocusedColor = CRGBA(150, 0, 192, 160),
	WanderDecalTexture = "r2_wander.tga",
	WanderDecalSize = 0.55,
	AutoGroupMaxDist = 4,
	-- display of inaccessible pos on map
	InaccessiblePosColor0			= CRGBA(255, 0, 0, 255),
	InaccessiblePosColor1			= CRGBA(200, 217, 0, 255),
	InaccessiblePosAnimDurationInMS = 500,
	DecalTopBlendStartDist = 3,		-- Distance in meters at which the color starts to fade for player & selection decals (at top)
	DecalBottomBlendStartDist = 1,  -- Distance in meters at which the color starts to fade for player & selection decals (at bottom)
	DecalBlendLength = 1.5			-- Length for decal color fading (player & selection decals)
}